tools/dumpkey/dumpkey.cpp

```diff
@@ -136,6 +147,56 @@ int main(int argc, const char **argv)
         }
         else
         {
+            Owned<IPropertyTree> metadata = index->getMetadata();
+            Owned<IOutputMetaData> diskmeta;
+            ArrayOf<const RtlFieldInfo *> fields;  // Note - the lifetime of the array needs to extend beyond the lifetime of outmeta. The fields themselves are shared with diskmeta, and do not need to be released.
+            Owned<IOutputMetaData> outmeta;
+            Owned<IHThorIndexReadArg> helper;
+            Owned<IXmlWriterExt> writer;
+            class MyIndexCallback : public CInterfaceOf<IThorIndexCallback>
+            {
+            public:
+                MyIndexCallback(IKeyCursor *_cursor) : cursor(_cursor) {}
+                virtual unsigned __int64 getFilePosition(const void * row)
+                {
+                    return cursor->getFPos();
+                }
+                virtual byte * lookupBlob(unsigned __int64 id)
+                {
+                    UNIMPLEMENTED;
+                }
+                Linked<IKeyCursor> cursor;
+            } callback(cursor);
+            unsigned count = globals->getPropInt("recs", 1);
+            const RtlRecordTypeInfo *outRecType = nullptr;
+            if (metadata && metadata->hasProp("_rtlType"))
+            {
+                MemoryBuffer layoutBin;
+                metadata->getPropBin("_rtlType", layoutBin);
+                diskmeta.setown(new CDeserializedOutputMetaData(layoutBin, true, &callback));
+                writer.setown(new SimpleOutputWriter);
+                if (globals->hasProp("fields"))
+                {
+                    StringArray fieldNames;
+                    fieldNames.appendList(globals->queryProp("fields"), ",");
+                    const RtlRecord &inrec = diskmeta->queryRecordAccessor(true);
+                    ForEachItemIn(idx, fieldNames)
+                    {
+                        unsigned fieldNum = inrec.getFieldNum(fieldNames.item(idx));
+                        if (fieldNum == (unsigned) -1)
+                            throw MakeStringException(0, "Requested output field '%s' not found", fieldNames.item(idx));
+                        fields.append(inrec.queryField(fieldNum));
+                    }
+                    fields.append(nullptr);
+                    outRecType = new RtlRecordTypeInfo(type_record, 0, fields.getArray(0)); // MORE - is length needed?
```

Comment:
length would ideally be the minimum length of the record.  It may cause problems in the future (for fixed size records) if not filled in.